### ISSUE TITLE
C Compiler Failed

### ISSUE THREAD

**Issue by Lvl4Sword** [bitcoin-wallet/bitcoin-wallet#485]
Running `gradle clean :native-scrypt:copy test build` from within the wallet folder on a Ubuntu 16.04 LTS install, I get the following:

```
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/2.10/userguide/gradle_daemon.html.
Incremental java compilation is an incubating feature.
:wallet:clean UP-TO-DATE
:native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC
/home/user/Desktop/bitcoin-wallet/native-scrypt/src/scrypt_jni.c:3:19: fatal error: errno.h: No such file or directory
 #include <errno.h>
                   ^
compilation terminated.

In file included from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/scrypt_platform.h:7:0,
                 from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/crypto_scrypt-nosse.c:29:
/home/user/Desktop/bitcoin-wallet/native-scrypt/src/config.h:9:24: fatal error: sys/limits.h: No such file or directory
 #include <sys/limits.h>
                        ^
compilation terminated.

In file included from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/scrypt_platform.h:7:0,
                 from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/sha256.c:26:
/home/user/Desktop/bitcoin-wallet/native-scrypt/src/config.h:9:24: fatal error: sys/limits.h: No such file or directory
 #include <sys/limits.h>
                        ^
compilation terminated.

:native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC'.
> Multiple build operations failed.
      C compiler failed while compiling scrypt_jni.c.
      C compiler failed while compiling crypto_scrypt-nosse.c.
      C compiler failed while compiling sha256.c.
  See the complete log at: file:///home/user/Desktop/bitcoin-wallet/native-scrypt/build/tmp/compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC/output.txt

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 24.525 secs
```

Please advise on how to continue.

**Comment by Lvl4Sword** [comment]
Running `gradle clean :native-scrypt:copy test build --stacktrace` gives me the following:

> To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/2.10/userguide/gradle_daemon.html.
> Incremental java compilation is an incubating feature.
> :wallet:clean UP-TO-DATE
> :native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC
> /home/user/Desktop/bitcoin-wallet/native-scrypt/src/scrypt_jni.c:3:19: fatal error: errno.h: No such file or directory
>  #include <errno.h>
>                    ^
> compilation terminated.
> 
> In file included from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/scrypt_platform.h:7:0,
>                  from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/crypto_scrypt-nosse.c:29:
> /home/user/Desktop/bitcoin-wallet/native-scrypt/src/config.h:9:24: fatal error: sys/limits.h: No such file or directory
>  #include <sys/limits.h>
>                         ^
> compilation terminated.
> 
> In file included from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/scrypt_platform.h:7:0,
>                  from /home/user/Desktop/bitcoin-wallet/native-scrypt/src/sha256.c:26:
> /home/user/Desktop/bitcoin-wallet/native-scrypt/src/config.h:9:24: fatal error: sys/limits.h: No such file or directory
>  #include <sys/limits.h>
>                         ^
> compilation terminated.
> 
> :native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC'.
> > Multiple build operations failed.
>       C compiler failed while compiling scrypt_jni.c.
>       C compiler failed while compiling crypto_scrypt-nosse.c.
>       C compiler failed while compiling sha256.c.
>   See the complete log at: file:///home/user/Desktop/bitcoin-wallet/native-scrypt/build/tmp/compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC/output.txt
> 
> * Try:
> Run with --info or --debug option to get more log output.
> 
> * Exception is:
> org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':native-scrypt:compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC'.
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
> 	at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
> 	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
> 	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
> 	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
> 	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
> 	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
> 	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
> 	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:203)
> 	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:185)
> 	at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:66)
> 	at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:50)
> 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:25)
> 	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:110)
> 	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
> 	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
> 	at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
> 	at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
> 	at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
> 	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
> 	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
> 	at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:154)
> 	at org.gradle.internal.Factories$1.create(Factories.java:22)
> 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
> 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:52)
> 	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:151)
> 	at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
> 	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:99)
> 	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:93)
> 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
> 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:62)
> 	at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:93)
> 	at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:82)
> 	at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:94)
> 	at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
> 	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
> 	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:43)
> 	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:28)
> 	at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:78)
> 	at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:48)
> 	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:52)
> 	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
> 	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
> 	at org.gradle.util.Swapper.swap(Swapper.java:38)
> 	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.health.DaemonHealthTracker.execute(DaemonHealthTracker.java:40)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:66)
> 	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
> 	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.health.HintGCAfterBuild.execute(HintGCAfterBuild.java:41)
> 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
> 	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
> 	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:246)
> 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
> 	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
> Caused by: org.gradle.internal.operations.MultipleBuildOperationFailures: Multiple build operations failed.
>     C compiler failed while compiling scrypt_jni.c.
>     C compiler failed while compiling crypto_scrypt-nosse.c.
>     C compiler failed while compiling sha256.c.
> See the complete log at: file:///home/user/Desktop/bitcoin-wallet/native-scrypt/build/tmp/compileScryptArmeabi-v7aDebugSharedLibraryScryptMainC/output.txt
> 	at org.gradle.internal.operations.DefaultBuildOperationQueue.waitForCompletion(DefaultBuildOperationQueue.java:72)
> 	at org.gradle.nativeplatform.toolchain.internal.NativeCompiler.execute(NativeCompiler.java:78)
> 	at org.gradle.nativeplatform.toolchain.internal.NativeCompiler.execute(NativeCompiler.java:43)
> 	at org.gradle.nativeplatform.toolchain.internal.OutputCleaningCompiler.compileSources(OutputCleaningCompiler.java:46)
> 	at org.gradle.nativeplatform.toolchain.internal.OutputCleaningCompiler.execute(OutputCleaningCompiler.java:38)
> 	at org.gradle.nativeplatform.toolchain.internal.OutputCleaningCompiler.execute(OutputCleaningCompiler.java:26)
> 	at org.gradle.language.nativeplatform.internal.incremental.IncrementalNativeCompiler.doCleanIncrementalCompile(IncrementalNativeCompiler.java:153)
> 	at org.gradle.language.nativeplatform.internal.incremental.IncrementalNativeCompiler.execute(IncrementalNativeCompiler.java:105)
> 	at org.gradle.language.nativeplatform.internal.incremental.IncrementalNativeCompiler.execute(IncrementalNativeCompiler.java:49)
> 	at org.gradle.nativeplatform.internal.BuildOperationLoggingCompilerDecorator.execute(BuildOperationLoggingCompilerDecorator.java:38)
> 	at org.gradle.nativeplatform.internal.BuildOperationLoggingCompilerDecorator.execute(BuildOperationLoggingCompilerDecorator.java:22)
> 	at org.gradle.language.nativeplatform.tasks.AbstractNativeCompileTask.doCompile(AbstractNativeCompileTask.java:109)
> 	at org.gradle.language.nativeplatform.tasks.AbstractNativeCompileTask.compile(AbstractNativeCompileTask.java:98)
> 	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
> 	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$IncrementalTaskAction.doExecute(AnnotationProcessingTaskFactory.java:244)
> 	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:220)
> 	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$IncrementalTaskAction.execute(AnnotationProcessingTaskFactory.java:231)
> 	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:209)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
> 	... 68 more
> Caused by: org.gradle.nativeplatform.toolchain.internal.CommandLineToolInvocationFailure: C compiler failed while compiling scrypt_jni.c.
> 	at org.gradle.nativeplatform.toolchain.internal.DefaultCommandLineToolInvocationWorker.execute(DefaultCommandLineToolInvocationWorker.java:83)
> 	at org.gradle.nativeplatform.toolchain.internal.DefaultCommandLineToolInvocationWorker.execute(DefaultCommandLineToolInvocationWorker.java:30)
> 	at org.gradle.internal.operations.DefaultBuildOperationQueue$OperationHolder.run(DefaultBuildOperationQueue.java:110)
> 	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:106)
> 	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
> 	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:76)
> 	at org.gradle.internal.concurrent.ExecutorPolicy$PropagateFailures.onExecute(ExecutorPolicy.java:76)
> 	... 1 more
> 
> 
> BUILD FAILED
> 
> Total time: 21.124 secs

**Comment by schildbach** [comment]
Do you have the NDK installed and the `ANDROID_NDK_HOME` environment variable set?

**Comment by Lvl4Sword** [comment]
I do. `export ANDROID_NDK_HOME=$HOME/Android/android-ndk-r16b` has been set

**Comment by Lvl4Sword** [comment]
Whether I'm in bitcoin-wallet or the wallet directory, I still get a C compiler error.

**Comment by Lvl4Sword** [comment]
`export` gives the following, I've changed the name of the NDK directory in the chance I may have typoed it, no dice. I've fixed the environment variable to show this change.
```
declare -x ANDROID_HOME="/home/user/Android/Sdk"
declare -x ANDROID_NDK_HOME="/home/user/Android/NDK"
declare -x CLUTTER_IM_MODULE="xim"
declare -x COMPIZ_BIN_PATH="/usr/bin/"
declare -x COMPIZ_CONFIG_PROFILE="ubuntu-lowgfx"
declare -x DBUS_SESSION_BUS_ADDRESS="unix:abstract=/tmp/dbus-HxODqen8T4"
declare -x DEFAULTS_PATH="/usr/share/gconf/ubuntu.default.path"
declare -x DESKTOP_SESSION="ubuntu"
declare -x DISPLAY=":0"
declare -x GDMSESSION="ubuntu"
declare -x GDM_LANG="en_US"
declare -x GNOME_DESKTOP_SESSION_ID="this-is-deprecated"
declare -x GNOME_KEYRING_CONTROL=""
declare -x GNOME_KEYRING_PID=""
declare -x GTK2_MODULES="overlay-scrollbar"
declare -x GTK_IM_MODULE="ibus"
declare -x GTK_MODULES="gail:atk-bridge:unity-gtk-module"
declare -x HOME="/home/user"
declare -x IM_CONFIG_PHASE="1"
declare -x INSTANCE=""
declare -x JOB="unity-settings-daemon"
declare -x LANG="en_US.UTF-8"
declare -x LANGUAGE="en_US"
declare -x LESSCLOSE="/usr/bin/lesspipe %s %s"
declare -x LESSOPEN="| /usr/bin/lesspipe %s"
declare -x LIBGL_ALWAYS_SOFTWARE="1"
declare -x LOGNAME="user"
declare -x LS_COLORS="rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:"
declare -x MANDATORY_PATH="/usr/share/gconf/ubuntu.mandatory.path"
declare -x OLDPWD="/home/user/Desktop"
declare -x PATH="/home/user/bin:/home/user/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/user/Android/Sdk/tools"
declare -x PWD="/home/user/Desktop/bitcoin-wallet"
declare -x QT4_IM_MODULE="xim"
declare -x QT_ACCESSIBILITY="1"
declare -x QT_IM_MODULE="ibus"
declare -x QT_LINUX_ACCESSIBILITY_ALWAYS_ON="1"
declare -x QT_QPA_PLATFORMTHEME="appmenu-qt5"
declare -x SESSION="ubuntu"
declare -x SESSIONTYPE="gnome-session"
declare -x SESSION_MANAGER="local/dev-vm:@/tmp/.ICE-unix/1308,unix/dev-vm:/tmp/.ICE-unix/1308"
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x SSH_AUTH_SOCK="/run/user/1000/keyring/ssh"
declare -x TERM="xterm-256color"
declare -x UPSTART_EVENTS="xsession started"
declare -x UPSTART_INSTANCE=""
declare -x UPSTART_JOB="unity7"
declare -x UPSTART_SESSION="unix:abstract=/com/ubuntu/upstart-session/1000/1012"
declare -x USER="user"
declare -x VTE_VERSION="4205"
declare -x WINDOWID="60817418"
declare -x XAUTHORITY="/home/user/.Xauthority"
declare -x XDG_CONFIG_DIRS="/etc/xdg/xdg-ubuntu:/usr/share/upstart/xdg:/etc/xdg"
declare -x XDG_CURRENT_DESKTOP="Unity"
declare -x XDG_DATA_DIRS="/usr/share/ubuntu:/usr/share/gnome:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop"
declare -x XDG_GREETER_DATA_DIR="/var/lib/lightdm-data/user"
declare -x XDG_MENU_PREFIX="gnome-"
declare -x XDG_RUNTIME_DIR="/run/user/1000"
declare -x XDG_SEAT="seat0"
declare -x XDG_SEAT_PATH="/org/freedesktop/DisplayManager/Seat0"
declare -x XDG_SESSION_DESKTOP="ubuntu"
declare -x XDG_SESSION_ID="c1"
declare -x XDG_SESSION_PATH="/org/freedesktop/DisplayManager/Session0"
declare -x XDG_SESSION_TYPE="x11"
declare -x XDG_VTNR="7"
declare -x XMODIFIERS="@im=ibus"
```

**Comment by Lvl4Sword** [comment]
@schildbach What system do you use to compile bitcoin-wallet? Perhaps there are packages I'm missing which aren't documented for this?

**Comment by DanAlbert** [comment]
I looks like the problem is that the build files do not support [unified headers](https://android.googlesource.com/platform/ndk/+/ndk-r15-release/docs/UnifiedHeaders.md). The deprecated headers were removed from NDK r16. I'm not entirely sure what the defaults were for the various gradle plugins though.

**Comment by schildbach** [comment]
Sorry, I wanted to get back to you much earlier. I'm using an almost pure Ubuntu 16.04, and a few days ago I also tried the Android NDK r16b which worked fine for me (was using r12b before).

**Comment by schildbach** [comment]
Do you have the package `build-essential` installed?

**Comment by DanAlbert** [comment]
Shouldn't be required. The NDK provides all the same packages.

**Comment by HashEngineering** [comment]
I have been using r12b also for several of my projects that are forked from bitcoin-wallet.  I haven't upgraded to r16 for any builds yet.

**Comment by Nutomic** [comment]
I can confirm that NDK r12b fixes the build issues for me (on latest Manjaro).

**Comment by schildbach** [comment]
Ok you were right, on my integration test machine the build fails with r16b too. r15c works though.

It seems as we need someone who is proficient in C++ and can fix the scrypt source to be compatible with newer NDK releases. We also have the problem that in August 2019, Google will require 64bit native libraries and as of now the 64bit build is broken (in a subtle way; yields wrong scrypt results).

**Comment by Lvl4Sword** [comment]
Confirmed that r15c works, while r16b fails.

**Comment by DanAlbert** [comment]
It's not actually a C++ issue, it's an issue with your build scripts. If I understand the situation correctly, the gradle plugin method of building NDK code does not automatically use the new (unified) headers like it ought to. Updating to the more modern `externalNativeBuild` (using cmake or ndk-build within gradle) would do that appropriately. Unfortunately my knowledge of the gradle side of things is limited, but hopefully that gets you pointed in the right direction.

**Comment by schildbach** [comment]
Fixed by #532.
